File: src/reorder_phylo.c
/* reorder_phylo.c       2021-04-07 */

/* Copyright 2008-2021 Emmanuel Paradis */

/* This file is part of the R-package `ape'. */
/* See the file ../COPYING for licensing issues. */

#include <R.h>

#define DO_NODE_PRUNING\
    /* go back down in `edge' to set `neworder' */\
    for (j = 0; j <= i; j++) {\
        /* if find the edge where `node' is */\
        /* the descendant, make as ready */\
        if (edge2[j] == node) ready[j] = 1;\
	if (edge1[j] != node) continue;\
	neworder[nextI] = j + 1;\
	ready[j] = 0; /* mark the edge as done */\
	nextI++;\
    }

void neworder_pruningwise(int *ntip, int *nnode, int *edge1,
			  int *edge2, int *nedge, int *neworder)
{
    int *ready, *Ndegr, i, j, node, nextI, n;

    nextI = *ntip +  *nnode;
    Ndegr = (int*)R_alloc(nextI, sizeof(int));
    memset(Ndegr, 0, nextI*sizeof(int));
    for (i = 0; i < *nedge; i++) (Ndegr[edge1[i] - 1])++;

    ready = (int*)R_alloc(*nedge, sizeof(int));

    /* `ready' indicates whether an edge is ready to be */
    /* collected; only the terminal edges are initially ready */
    for (i = 0; i < *nedge; i++)
	    ready[i] = (edge2[i] <= *ntip) ? 1 : 0;

    /* `n' counts the number of times a node has been seen. */
    /* This algo will work if the tree is in cladewise order, */
    /* so that the nodes of "cherries" will be contiguous in `edge'. */
    n = 0;
    nextI = 0;
    while (nextI < *nedge - Ndegr[*ntip]) {
        for (i = 0; i < *nedge; i++) {
            if (!ready[i]) continue;
	    if (!n) {
	        /* if found an edge ready, initialize `node' and start counting */
	        node = edge1[i];
		n = 1;
	    } else { /* else counting has already started */
	        if (edge1[i] == node) n++;
		else {
		    /* if the node has changed we checked that all edges */
		    /* from `node' have been found */
		    if (n == Ndegr[node - 1]) {
		        DO_NODE_PRUNING
		    }
		    /* in all cases reset `n' and `node' and carry on */
		    node = edge1[i];
		    n = 1;
		}
	    } /* go to the next edge */
	    /* if at the end of `edge', check that we can't do a node */
	    if (n == Ndegr[node - 1]) {
	        DO_NODE_PRUNING
		n = 0;
	    }
        }
    }
    for (i = 0; i < *nedge; i++) {
        if (!ready[i]) continue;
	neworder[nextI] = i + 1;
	nextI++;
    }
}

File: R/reorder.phylo.R
## reorder.phylo.R (2017-07-28)

##   Internal Reordering of Trees

## Copyright 2006-2017 Emmanuel Paradis, 2017 Klaus Schliep

## This file is part of the R-package `ape'.
## See the file ../COPYING for licensing issues.

.reorder_ape <- function(x, order, index.only, nb.tip, io)
{
    nb.edge <- dim(x$edge)[1]
    if (!is.null(attr(x, "order")))
        if (attr(x, "order") == order)
            if (index.only) return(1:nb.edge) else return(x)
    nb.node <- x$Nnode
    if (nb.node == 1)
        if (index.only) return(1:nb.edge) else return(x)

    if (io == 3) {
        x <- reorder(x)
        neworder <-
            .C(neworder_pruningwise, as.integer(nb.tip),
               as.integer(nb.node), as.integer(x$edge[, 1]),
               as.integer(x$edge[, 2]), as.integer(nb.edge),
               integer(nb.edge))[[6]]
    } else {
        neworder <- reorderRcpp(x$edge, as.integer(nb.tip),
                                as.integer(nb.tip + 1L), io)
    }
    if (index.only) return(neworder)
    x$edge <- x$edge[neworder, ]
    if (!is.null(x$edge.length))
        x$edge.length <- x$edge.length[neworder]
    attr(x, "order") <- order
    x
}

reorder.phylo <- function(x, order = "cladewise", index.only = FALSE, ...)
{
    ORDER <- c("cladewise", "postorder", "pruningwise")
    io <- pmatch(order, ORDER)
    if (is.na(io)) stop("ambiguous order")
    order <- ORDER[io]
    .reorder_ape(x, order, index.only, length(x$tip.label), io)
}

reorder.multiPhylo <- function(x, order = "cladewise", ...)
{
    ORDER <- c("cladewise", "postorder", "pruningwise")
    io <- pmatch(order, ORDER)
    if (is.na(io)) stop("ambiguous order")
    order <- ORDER[io]
    oc <- oldClass(x)
    class(x) <- NULL
    labs <- attr(x, "TipLabel")
    x <-
        if (is.null(labs)) lapply(x, reorder.phylo, order = order)
        else lapply(x, .reorder_ape, order = order, index.only = FALSE,
                    nb.tip = length(labs), io = io)
    if (!is.null(labs)) attr(x, "TipLabel") <- labs
    class(x) <- oc
    x
}

cladewise <- function(x) reorder(x, "cladewise", index.only = TRUE)

postorder <- function(x) reorder(x, "postorder", index.only = TRUE)

rotateConstr <- function(phy, constraint)
{
    D <- match(phy$tip.label, constraint)
    n <- Ntip(phy)

    P <- c(as.list(1:n), prop.part(phy))
    e1 <- phy$edge[, 1L]
    e2 <- phy$edge[, 2L]

    foo <- function(node) {
        i <- which(e1 == node) # the edges where 'node' is ancestral
        desc <- e2[i] # the descendants of 'node'
        ## below, min() seems to work better than median() which
        ## seems to work better than mean() which seems to work
        ## better than sum()
        o <- order(sapply(desc, function(x) min(D[P[[x]]])))
        for (k in o) {
            j <<- j + 1L
            neworder[j] <<- i[k]
            if ((dk <- desc[k]) > n) foo(dk)
        }
    }

    neworder <- integer(Nedge(phy))
    j <- 0L
    foo(n + 1L)
    phy$edge <- phy$edge[neworder, ]
    if (!is.null(phy$edge.length))
        phy$edge.length <- phy$edge.length[neworder]
    attr(phy, "order") <- "cladewise"
    phy
}
